### PR TITLE
OPS: change license for reproducible builds

### PR DESCRIPTION
**Pull request by winterrdog**
This replaces the old GPL-licensed `apkdiff` utility, originally taken from
Signal, with a full reimplementation from scratch

The new implementation keeps the same purpose and functionality, but uses
different algorithms and data structures throughout. For instance, ZIP
entries are now indexed by filename for direct matching instead of repeatedly
searching through the entry lists. XML files are compared using an iterative
stack-based tree traversal which in turn allowed for a higher stack limit
(might even be removed in the future), grouping children by tag before
comparing them. The `resources.arsc` comparison also uses a unified
type-based comparison approach rather than the original per-type branching.
Just to name a few.

This implementation does not carry over the original GPL code and can
therefore be licensed under MIT. 

> [!NOTE]
>  Instructions on how to verify that builds are reproducible are in the `README.md`
> in the changes.

### CODE REVIEW

File: reproducible-builds/apkdiff/pyproject.toml [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+name = "apk-compare"
 version = "0.1.0"
-description = "APK comparison utility"
+description = "A utility that compares APKs."
 requires-python = ">=3.12"
 dependencies = [
     "androguard>=4.1.3",
     "tqdm>=4.67.1",
     "loguru>=0.7.3",
     "fonttools==4.63.0",
```
Comment (Overtorment) [inline reproducible-builds/apkdiff/pyproject.toml]: since it a rewrite, can we drop  the useless depps? some dep scanning services scream that there are vulns there. annoying

Comment (winterrdog) [reply reproducible-builds/apkdiff/pyproject.toml]: Sure thing!
Going to remove them when I retouch

Comment (winterrdog) [reply reproducible-builds/apkdiff/pyproject.toml]: pushed 3b456af1492efca783a313692f6bca66b10ebaab

only retained `androguard` and `loguru` because:
- `androguard`: used to parse APK internals (`AndroidManifest.xml`, `resources.arsc`, binary XML files). it covers most of the core functionality
- `loguru`: used for _structured_ logging with formatting, timestamps, and suppresses noisy output from `androguard`/`deepdiff` libraries

@Overtorment  is that fine ?

File: reproducible-builds/apkdiff/apkdiff.py [hunk 2]
```diff
@@ -252,180 +316,192 @@ def compare_resources_arsc(first_entry_bytes: bytes, second_entry_bytes: bytes)
...
-
-        # Compare matching elements recursively
-        for child1, child2 in zip(list1, list2):
-            differences.extend(compare_xml_elements(child1, child2, current_path))
 
-    return differences
+def compare(apk1, apk2) -> bool:
+    logger.info("Comparing: \n\t{}\n\t{}\n", apk1, apk2)
+    with open_apk(apk1) as zip1, open_apk(apk2) as zip2:
+        return cmp_entry_names(zip1, zip2) and cmp_entry_contents(zip1, zip2)
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:500]: **[HIGH]** `and` short-circuits—name mismatch skips content comparison entirely. Master always ran both. Your mismatch forensics just got vaporized mid-chamber.

File: reproducible-builds/apkdiff/tests/test_apkdiff.py [hunk 3]
```diff
@@ -0,0 +1 @@
+# TODO: add tests to avoid regressions in apkdiff.py after
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/tests/test_apkdiff.py]: **[HIGH]** One-line TODO where regression tests belong. You rewrote the entire comparator and left `apkdiff.py` scientifically unobserved. CONTRIBUTING wanted happy paths and edge cases—not wishful thinking.

File: reproducible-builds/apkdiff/apkdiff.py [hunk 4]
```diff
@@ -53,101 +37,169 @@ class XmlDifference:
...
+    path: str
+    attr_name: Optional[str] = None
+    first_val: Optional[str] = None
+    second_val: Optional[str] = None
+    child_tag: Optional[str] = None
+    first_element_attrs: Optional[dict] = None
+    second_element_attrs: Optional[dict] = None
+
+
+class LoguruInterceptHandler(logging.Handler):
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:55]: **[HIGH]** `LoguruInterceptHandler` plus forced `basicConfig` to hush two libraries? Old code: two lines. This is Aperture-grade bureaucracy for a CLI that compares zip files.

File: reproducible-builds/apkdiff/util.py [hunk 5]
```diff
@@ -1,198 +1,189 @@
...
+        exclude_attrs = set(exclude_attrs) | default_exclude_attrs
+
+    def safe_sort(iterable):
+        try:
+            return sorted(iterable)
+        except TypeError:
+            return list(iterable)
+
+    # explicit stack with tuples of: (obj1, obj2, path, current_depth)
+    stack = [(obj1, obj2, path, curr_depth)]
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/util.py:65]: **[HIGH]** Recursive `deep_compare` became a manual stack machine—for identical behavior. Same bloat infects `show_diffs` and `cmp_xml_elements`. MIT license, maximal diff.

File: reproducible-builds/apkdiff/apkdiff.py [hunk 6]
```diff
@@ -160,90 +212,102 @@ def handle_special_cases(filename: str, bytes1: bytes, bytes2: bytes):
...
+    allowed_paths = set(ALLOWED_ARSC_DIFF_PATHS)
     total_diffs = defaultdict(list)
-
     success = True
 
     for package_name in all_package_names:
         # Check if package exists in both files
         if package_name not in first_arsc.packages:
-            print(f"Package only in source file: {package_name}")
+            logger.info("Package only in target file: {}", package_name)
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:306]: **[MEDIUM]** Package-missing logs swapped: first APK absence screams 'target file', second screams 'source'. Debug output now lies. Enjoy decoding that cake.

File: reproducible-builds/README.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+`apkdiff.py` compares APK files and ignores expected differences:
+
+- App signing metadata (certificates, signatures)
+- Bugsnag `BUILD_UUID` values
+- Play Store bundle artifacts
+
+It performs byte-by-byte comparison of all other files, with special handling for Android binary formats (manifests, resources). See [`apkdiff/apkdiff.py`](apkdiff/apkdiff.py) for implementation.
+
+<!--
```
Comment (GladosBlueWallet) [inline reproducible-builds/README.md:55]: **[MEDIUM]** Pytest wired in `pyproject.toml`, test docs HTML-commented with TODO. You built the test chamber, hid the door, and filed it under 'later'.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
License-clean refactor of the reproducible-build `apkdiff` tool—drop `tqdm`, reimplement comparators, add README. Noble. Shame the subject refuses to prove it works: zero `apkdiff.py` tests after gutting GPL code, `compare()` now short-circuits like a lazy test subject, and you bolted a Loguru interception rig where two log lines once sufficed.

### Inline findings (could not anchor on diff)

- `CONTRIBUTING.md:3` — **[LOW]** Four of five commits ignore mandated prefixes (`REF:`, `DOC:`, `TST:`). Only the `tqdm` drop got `FIX:`. Procedure exists; compliance optional, apparently.

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ojokne] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ojokne

File: reproducible-builds/apkdiff/pyproject.toml (hunk 1 above)
Comment (Overtorment) [reply reproducible-builds/apkdiff/pyproject.toml]: lets merge, will see after if aikido/socket will complain

**Comment by Overtorment** [comment]
glados complains look valid

**Comment by winterrdog** [comment]
> glados complains look valid

Yes! Going to address them 👍

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ojokne] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ojokne

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ojokne] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ojokne

File: reproducible-builds/apkdiff/apkdiff.py (hunk 6 above)
Comment (winterrdog) [reply reproducible-builds/apkdiff/apkdiff.py:306]: intentional

File: reproducible-builds/README.md (hunk 7 above)
Comment (winterrdog) [reply reproducible-builds/README.md:55]: intentional, as well. i will work on them in a separate PR

File: reproducible-builds/apkdiff/apkdiff.py (hunk 2 above)
Comment (winterrdog) [reply reproducible-builds/apkdiff/apkdiff.py:500]: yes, intentional! here, we aim for a strict and complete APK matching for 100% reproducibility

File: reproducible-builds/apkdiff/tests/test_apkdiff.py (hunk 3 above)
Comment (winterrdog) [reply reproducible-builds/apkdiff/tests/test_apkdiff.py]: fixed

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @ojokne] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/ojokne

File: reproducible-builds/apkdiff/apkdiff.py [hunk 8]
```diff
@@ -53,101 +37,169 @@ class XmlDifference:
     "stamp-cert-sha256",
 ]
```
Comment (ojokne) [inline reproducible-builds/apkdiff/apkdiff.py:38]: I think it would be more transparent if we also clearly display to the user that the following files are ignored when doing the comparsion just like you clearly indicated for Bugsnag BUILD_UUID on line 253. What do you think?

Comment (winterrdog) [reply reproducible-builds/apkdiff/apkdiff.py:38]: done

File: reproducible-builds/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
+# Reproducible Builds
+
+Reproducible builds for BlueWallet. Build the same APK twice and verify they're byte-for-byte identical.
```
Comment (ojokne) [inline reproducible-builds/README.md:3]: The overall aim is for users to compare their local build against the official released version on playstore, Github, FDroid etc, this documentation is very helpful for someone testing and reviewing this pr so i think i would be better to move it to the pr description.

After this PR is merged, we shall be left with configuring the Github actions to build a reproducible apk then we can update the docs with how to verify the locally built apk with the official release.

What do you think about this?

**Review by ojokne — APPROVED** [review]
Tested with several pairs of apks,

Clearly indicates the files that are ignored

Was able to show when apks matched and when they did not match

LGTM

**Review by Overtorment — APPROVED** [review]
lgtm

File: reproducible-builds/apkdiff/tests/test_apkdiff.py (hunk 3 above)
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/tests/test_apkdiff.py:1]: **[HIGH]** A file whose only payload is a TODO. You reimplemented the entire APK comparator and left zero regression tests for manifest/ARSC/XML paths — CONTRIBUTING asked for real tests, not decorative filenames.

File: reproducible-builds/apkdiff/apkdiff.py (hunk 4 above)
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:55]: **[HIGH]** LoguruInterceptHandler plus forced basicConfig — thirty lines to hush libraries when `logger.disable('androguard')` worked fine. Over-engineered noise for a CLI diff tool.

File: reproducible-builds/apkdiff/util.py [hunk 10]
```diff
@@ -1,198 +1,189 @@
...
+                formatted = indented + field + " =>"
+                result.append(formatted)
+                stack.append((data, indent + 2))
+            else:
+                formatted = f"{indented}{field} => {data}"
+                result.append(formatted)
+    return "\n".join(result)
 
 
 def deep_compare(
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/util.py:25]: **[HIGH]** 189-line iterative `deep_compare` with dot-notation heuristics and stack machinery replacing a simpler recursive version — bloat cosplaying as a license refactor.

File: reproducible-builds/apkdiff/apkdiff.py [hunk 11]
```diff
@@ -252,180 +316,197 @@ def compare_resources_arsc(first_entry_bytes: bytes, second_entry_bytes: bytes)
...
-    return differences
+def compare(apk1, apk2) -> bool:
+
+    def display_ignored_files():
+        return ", ".join(map(lambda x: f"\n\t\t\t'{x}'", IGNORE_FILES))
+
+    logger.info("Comparing: \n\t\t\t{}\n\t\t\t{}\n", apk1, apk2)
+    logger.info("Files ignored: {}.", display_ignored_files())
+    with open_apk(apk1) as zip1, open_apk(apk2) as zip2:
+        return cmp_entry_names(zip1, zip2) and cmp_entry_contents(zip1, zip2)
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:500]: **[MEDIUM]** `cmp_entry_names(...) and cmp_entry_contents(...)` short-circuits; master always ran both. Name mismatches now skip byte comparison and never extract `mismatches/` — worse forensics, same boolean, less data for humans.

File: reproducible-builds/apkdiff/apkdiff.py [hunk 12]
```diff
@@ -53,101 +37,169 @@ class XmlDifference:
...
+
+# Force-route Python's standard logging through Loguru
+logging.basicConfig(
+    handlers=[LoguruInterceptHandler()],
+    level=logging.DEBUG,
+    force=True,
+)
+
+# Shut-up noisy libraries
+logging.getLogger("deepdiff").setLevel(logging.ERROR)
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:76]: **[MEDIUM]** Silencing `deepdiff` logger — dependency that isn't in `pyproject.toml`. Dead config from a library you don't ship.

File: reproducible-builds/apkdiff/util.py [hunk 13]
```diff
@@ -1,198 +1,189 @@
...
+            for key in common_pairs:
+                # use dot notation for string keys after a numeric list/tuple index
+                use_dot = False
+                if isinstance(key, str) and p and p[-1] == "]":
+                    # check if the content before the final ] is a
+                    # digit (list index)
+                    try:
+                        bracket_start = p.rfind("[")
+                        if bracket_start < 0:
+                            continue
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/util.py:147]: **[LOW]** `continue` on `bracket_start < 0` skips comparing a dict key entirely instead of falling back — silent false-negative if that branch ever fires.

File: reproducible-builds/apkdiff/apkdiff.py [hunk 14]
```diff
@@ -252,180 +316,197 @@ def compare_resources_arsc(first_entry_bytes: bytes, second_entry_bytes: bytes)
...
         # Compare each package element
-        for i in tqdm(range(len(packages1))):
-            pkg1 = packages1[i]
-            pkg2 = packages2[i]
+        total = len(packages1)
+        log_interval = max(1, total // 10)  # Log at 10% intervals, minimum every item
+        for idx, (pkg1, pkg2) in enumerate(zip(packages1, packages2)):
+            if idx % log_interval == 0 or idx == total - 1:
+                progress_pct = ((idx + 1) * 100) // total
+                print(
```
Comment (GladosBlueWallet) [inline reproducible-builds/apkdiff/apkdiff.py:331]: **[LOW]** `print(..., end='\r')` progress bar while everything else uses Loguru — two logging systems in one script, because one annoyance wasn't enough.

**Review by GladosBlueWallet — CHANGES_REQUESTED** [review]
MIT/GPL refactor of `apkdiff` — license cleanup, drop `tqdm`/`fonttools`, Loguru logging. Core compare logic mostly intact, but you shipped a 500-line tool rewrite with a TODO where tests belong, a logging cathedral, and a short-circuit that kills diagnostics. Science progresses; your test chamber does not.

### Inline findings (could not anchor on diff)

- `reproducible-builds/apkdiff/README.md:55` — **[MEDIUM]** Test instructions HTML-commented out while `test_utils.py` ships and `test_apkdiff.py` is empty — onboarding docs lie about what exists.
- `reproducible-builds/apkdiff/uv.lock:1` — **[MEDIUM]** `pytest` declared under `[project.optional-dependencies]` dev but absent from lock — `uv sync` won't install the test runner your README pretends you'll use.
